File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        ## Creating adjanceny and indegree
        adj = {}
        indegree = [0] * numCourses

        for next, prev in prerequisites:
            if prev not in adj:
                adj[prev] = [next]
            else:
                adj[prev].append(next)
            indegree[next] += 1
        
        # BFS queue

        queue = deque()
        for node in range(numCourses):
            if indegree[node] == 0:
                queue.append(node)
        
        # BFS
        node_visited = 0
        while queue:
            node = queue.popleft()
            node_visited += 1
            for neighbour in adj.get(node,[]):
                indegree[neighbour] -= 1
                if indegree[neighbour] == 0:
                    queue.append(neighbour)
        return node_visited == numCourses

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int [] indegree = new int[numCourses];
        HashMap<Integer, List<Integer>> preMap = new HashMap<>();
        HashSet<Integer> seen = new HashSet<>();
        for(int i = 0; i < numCourses; i++){
            preMap.put(i, new ArrayList<>());
        }
        for(int [] edges : prerequisites){
            int u = edges[0];
            int v = edges[1];
            indegree[u] += 1;
            preMap.get(v).add(u);
        }
        Queue<Integer> queue = new LinkedList<>();
        for(int i = 0; i < numCourses; i++){
            if (indegree[i] == 0){
                queue.add(i);
            }
        }
        while(!queue.isEmpty()){
            int node = queue.remove();
            seen.add(node);
            List<Integer> neiNodes = preMap.get(node);
            for(int neinode : neiNodes){
                if (indegree[neinode] - 1 == 0){
                    queue.add(neinode);
                }
                indegree[neinode] -= 1;
            }
        }
        return seen.size() == numCourses;

    }
}